solutions/cf-609E.mdx

```diff
@@ -0,0 +1,179 @@
+---
+id: cf-609E
+source: CF
+title: Minimum spanning tree for each edge
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/22187)
+
+## Implementation 
+
+As mentioned in the official editorial, the answer for each edge is removing the heaviest edge in the MST in
+the path from vertex $u$ to $v$ and add the given weight. This will ensure that the MST is still connected.
+
+**Time Complexity:** $\mathcal{O}(M\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+const int MAXN = 2e5 + 1;
+const int MAXL = 20; // approximately maximum log
+
+// for euler-tour
+array<int, MAXN> enter_time, exit_time, depth;
+
+// for binary lifting
+array<array<int, MAXL>, MAXN> up, max_up;
+// up[i][j] stores the 2^jth ancestor of vertex i
+// max_up[i][j] stores the maximum weight of all the edges
+// from i to its 2^jth ancestor
+
+map<pair<int, int>, int> weight; // weight of edge {i, j}
+vector<int> graph[MAXN]; // adjacency list for the original MST
+int timer = 1;
+int n, m;
+
+// here we build the MST using kruskals' algorithm
+// returns the sum of the weights that took to build the MST
+ll kruskal(vector<array<int,4>>& edges){
+	
+	// sort by non-decreasing weight
+	sort(edges.begin(), edges.end(), [&](auto a, auto b){
+		return a[3] < b[3];
+	});
+	
+	// implementation of basic Disjoint Set Union
+	vector<int> parent(n+1), sz(n+1, 1);
+	iota(parent.begin(), parent.end(), 0);
```

Comment:
iota? if this does what i think it does, c++ already fills the vector with 0

Reply:
search up the cpp docs